libs/core/tests/integration/state_machine_tests.cpp

```diff
@@ -0,0 +1,177 @@
+  }
+
+  void TearDown() override
+  {
+    reactor_.Stop();
+  }
+
+  State OnA()
+  {
+    state_seen_ = static_cast<uint8_t>(State::A);
+    return State::B;
+  }
+
+  State OnB()
+  {
+    state_seen_ = static_cast<uint8_t>(State::B);
+    return State::C;
+  }
+
+  State OnC()
+  {
+    state_seen_ = static_cast<uint8_t>(State::C);
+    return State::A;
+  }
+
+  State OnSlowC()
+  {
+    state_seen_ = static_cast<uint8_t>(State::C);
+    std::this_thread::sleep_for(std::chrono::milliseconds(reactor_.ExecutionTooLongMs()));
+    return State::A;
+  }
+
+  State OnVerySlowC()
+  {
+    state_seen_ = static_cast<uint8_t>(State::C);
+    std::this_thread::sleep_for(std::chrono::milliseconds(reactor_.ThreadWatcherCheckMs() * 3));
+    return State::A;
+  }
+
+  static char const *ToString(State state)
+  {
+    char const *text = "unknown";
+
+    switch (state)
+    {
+    case State::A:
+      text = "A";
+      break;
+    case State::B:
+      text = "B";
+      break;
+    case State::C:
+      text = "C";
+      break;
+    }
+
+    return text;
+  }
+
+protected:
+  StateMachinePtr      state_machine_;
+  core::Reactor        reactor_;
+  std::atomic<uint8_t> state_seen_{std::numeric_limits<uint8_t>::max()};
+};
+
+// Basic test - does the reactor drive the state machine through all states
+TEST_F(ReactorTests, ReactorPassesThroughStates)
+{
+  // Note: because it is a fixture you need to upcast the this pointer
+  // clang-format off
+  state_machine_->RegisterHandler(State::A, static_cast<ReactorTests *>(this), &ReactorTests::OnA);
+  state_machine_->RegisterHandler(State::B, static_cast<ReactorTests *>(this), &ReactorTests::OnB);
+  state_machine_->RegisterHandler(State::C, static_cast<ReactorTests *>(this), &ReactorTests::OnC);
+  // clang-format on
+
+  reactor_.Attach(state_machine_);
+  reactor_.Start();
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(5));
+
+  EXPECT_NE(state_seen_, std::numeric_limits<uint8_t>::max());
+  EXPECT_EQ(reactor_.ExecutionsTooLongCounter(), 0);
+  EXPECT_EQ(reactor_.ExecutionsWayTooLongCounter(), 0);
+}
+
+// Test the state machine registers states that are taking too long
+TEST_F(ReactorTests, ReactorNoticesTooLongStates)
+{
+
+  // Note: because it is a fixture you need to upcast the this pointer
+  // clang-format off
+  state_machine_->RegisterHandler(State::A, static_cast<ReactorTests *>(this), &ReactorTests::OnA);
+  state_machine_->RegisterHandler(State::B, static_cast<ReactorTests *>(this), &ReactorTests::OnB);
+  state_machine_->RegisterHandler(State::C, static_cast<ReactorTests *>(this), &ReactorTests::OnSlowC);
+  // clang-format on
+
+  reactor_.Attach(state_machine_);
+  reactor_.Start();
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(reactor_.ExecutionTooLongMs() * 10));
+
+  EXPECT_NE(state_seen_, std::numeric_limits<uint8_t>::max());
+  EXPECT_NE(reactor_.ExecutionsTooLongCounter(), 0);
```

Comment:
Can it be expected to equal 10 without a risk of being flaky?

Reply:
Yeah, this should be half a second which should basically guarantee the reactor executes. If it doesn't that would indicate something has gone badly wrong